src/pybind/mgr/dashboard/services/auth/auth.py

```diff
@@ -25,6 +30,74 @@
 })
 
 
+class AuthType(str, Enum):
+    LOCAL = 'local'
+    SAML2 = 'saml2'
+    OAUTH2 = 'oauth2'
+
+
+class BaseAuth(abc.ABC):
+    LOGIN_URL: str
+    LOGOUT_URL: str
+    sso: bool
+
+    @staticmethod
+    def from_protocol(protocol: AuthType) -> Type["BaseAuth"]:
+        for subclass in BaseAuth.__subclasses__():
+            if subclass.__name__.lower() == protocol:
+                return subclass
+            for subsubclass in subclass.__subclasses__():
+                if subsubclass.__name__.lower() == protocol:
+                    return subsubclass
+        raise ValueError(f"Unknown auth backend: '{protocol}'")
+
+    @classmethod
+    def from_db(cls, db: Optional['SsoDB'] = None) -> Type["BaseAuth"]:
+        if db is None:
+            protocol = mgr.SSO_DB.protocol
+        else:
+            protocol = db.protocol
+        return cls.from_protocol(protocol)
+
+    class Config(TypedDict):  # pylint: disable=inherit-non-class
+        pass
+
+    @abc.abstractmethod
+    def to_dict(self) -> 'Config':
+        pass
+
+    @classmethod
+    @abc.abstractmethod
+    def from_dict(cls, s_dict) -> 'BaseAuth':
+        pass
+
+    @classmethod
+    def get_auth_name(cls):
+        return cls.__name__.lower()
+
+
+class Local(BaseAuth):
+    LOGIN_URL = '#/login'
+    LOGOUT_URL = '#/login'
+    sso = False
+
+    @classmethod
+    def get_auth_name(cls):
+        return cls.__name__.lower()
+
+    def to_dict(self) -> 'BaseAuth.Config':
+        return BaseAuth.Config()
+
+    @classmethod
+    def from_dict(cls, s_dict: BaseAuth.Config) -> 'Local':
+        # pylint: disable=unused-argument
+        return cls()
```

Comment:
If an abstract method in a derived class doesn't make sense (not ideal as it breaks the Liskov substitution principle), feel free to act accordingly:
```suggestion
    @classmethod
    def from_dict(cls, s_dict: BaseAuth.Config) -> 'Local':
        raise NotImplementedError
```

Reply:
I guess this might be a design problem more than a implemntation one as it doesn't make sense to have this methods for the `Local` and `Oauth2` but it does for `Saml2`

And it can not raise a `NotImplementedError` because on `load_sso_db`  the `from_dict` method of the active protocol will be called. [This is what I mean](https://github.com/ceph/ceph/blob/c37a00af2113d089d40b9d6df347fcb940efa41a/src/pybind/mgr/dashboard/services/sso.py#L82)
 
 What do you think @epuertat ?

Reply:
It's ok then!